src/chunk.c

```diff
@@ -1091,29 +1091,90 @@ chunk_create_from_hypercube_and_table_after_lock(const Hypertable *ht, Hypercube
 
 	Assert(OidIsValid(chunk_table_relid));
 	Assert(OidIsValid(current_chunk_schemaid));
+	Assert(OidIsValid(ht->main_table_relid));
+
+	Relation chunk_rel = table_open(chunk_table_relid, AccessShareLock);
+	Relation ht_rel = table_open(ht->main_table_relid, AccessShareLock);
+	TupleDesc tupdesc = RelationGetDescr(chunk_rel);
+
+	for (int attno = 0; attno < tupdesc->natts; attno++)
+	{
+		Form_pg_attribute att = TupleDescAttr(tupdesc, attno);
+		AttrNumber ht_attnum = InvalidAttrNumber;
+
+		/* Ignore dropped */
+		if (att->attisdropped)
+			continue;
+
+		ht_attnum = get_attnum(ht->main_table_relid, NameStr(att->attname));
+		/* Try to find the column in parent (matching on column name) */
+		if (ht_attnum == InvalidAttrNumber)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATATYPE_MISMATCH),
+					 errmsg("table \"%s\" contains column \"%s\" not found in parent \"%s\"",
+							RelationGetRelationName(chunk_rel),
+							NameStr(att->attname),
+							RelationGetRelationName(ht_rel)),
+					 errdetail("The new chunk can contain only the columns present in parent.")));
+
+		/*
+		 * PG16 and later does not allow generated columns on child tables if the parent
+		 * column is not generated. This is a change from PG15 and earlier, where the
+		 * child column could be generated even if the parent was not.
+		 * We check if a generated column is also generated in the parent here to disallow
+		 * this behavior in PG15 too.
+		 *
+		 * The case when the parent column is generated and the child column is not is handled
+		 * by Postgres code, which will throw an error.
+		 *
+		 * This check can be removed once we drop support for PG15.
+		 */
+		if (att->attgenerated && !get_attgenerated(ht->main_table_relid, ht_attnum))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_DATATYPE_MISMATCH),
+					 errmsg("column \"%s\" in chunk table must not be a generated column",
+							NameStr(att->attname)),
+					 errdetail("Chunk column must be generated if and only if parent column is "
+							   "also generated")));
+		}
+
+		/* Check that the chunk column has the same expression as the hypertable column */
+		if (att->attgenerated && get_attgenerated(ht->main_table_relid, ht_attnum))
+		{
+			char *chunk_expr = ts_get_attr_expr(chunk_rel, attno + 1);
+			char *ht_expr = ts_get_attr_expr(ht_rel, ht_attnum);
+
+			if (strcmp(chunk_expr, ht_expr) != 0)
+				ereport(ERROR,
+						(errcode(ERRCODE_DATATYPE_MISMATCH),
+						 errmsg("Chunk column \"%s\" must have the same expression as the "
+								"hypertable column.",
+								NameStr(att->attname))));
+		}
+	}
+	table_close(ht_rel, AccessShareLock);
```

Comment:
It is prudent to not release the lock and let the transaction commit release it (that is, use `NoLock`). Otherwise you can, in theory, have a modification of the hypertable definition after this lock release and have strange things happening at transaction commit.